### PR TITLE
AlertPulse -- checkpoint

### PR DESCRIPTION
**Pull request by gcmcnutt**
ok, gui prototype is in place -- and this works for bluetooth ch7 at least.

### CODE REVIEW

File: firmware/src/src/sense.cpp [hunk 1]
```diff
@@ -433,13 +433,33 @@ void calculate_Thread()
...
         int panch = trkset.panCh();
+        int alertch = trkset.alertCh();
         if(tltch > 0)
             channel_data[tltch - 1] = trpOutputEnabled == true ? tiltout_ui : trkset.Tlt_cnt();
         if(rllch > 0)
             channel_data[rllch - 1] = trpOutputEnabled == true ? rollout_ui : trkset.Rll_cnt();
         if(panch > 0)
             channel_data[panch - 1] = trpOutputEnabled == true ? panout_ui : trkset.Pan_cnt();
 
+        // TODO what is trpOutputEnabled?
```
Comment (gcmcnutt) [inline firmware/src/src/sense.cpp:444]: not sure if trpOutputEnabled will matter for alert pulse behavior

File: gui/src/boardbno055.cpp [hunk 2]
```diff
@@ -188,7 +189,7 @@ void BoardBNO055::parseIncomingHT(QString cmd)
     // Setting Data
     else if(cmd.left(5) == "$SET$") {
         QStringList setd = cmd.right(cmd.length()-5).split(',',Qt::KeepEmptyParts);
-        if(setd.length() == 22) {
+        if(setd.length() == 23) {
```
Comment (gcmcnutt) [inline gui/src/boardbno055.cpp:192]: i don't quite understand how this field count works here when i add a new field.  please review and let me know

File: gui/src/mainwindow.cpp [hunk 3]
```diff
@@ -768,6 +773,7 @@ void MainWindow::updateFromUI()
     pins[PIN_PPMIN] = ppin_index==1?-1:ppin_index;
     int but_index = ui->cmbButtonPin->currentIndex()+1;
     pins[PIN_BUTRESET] = but_index==1?-1:but_index;
+    // TODO something to do here for ALERT_CH?
```
Comment (gcmcnutt) [inline gui/src/mainwindow.cpp:776]: similarly here -- do I need to do similar treatment for alertch behavior?

**Comment by gcmcnutt** [comment]
so other things:
- what is the workflow for updating screen shots for README.md, docs, etc. -- I changed fields in the general tab of the gui
- what is process for updating online docs/help to explain this new field?
- when it comes time to release, i see there are DBs of releases, etc. -- for now, perhaps @dlktdr can do this

**Comment by gcmcnutt** [comment]
(also I didn't base this on top of the most recent fix as that seemed to have a couple of regressions on my tests)

File: gui/src/boardbno055.cpp (hunk 2 above)
Comment (dlktdr) [reply gui/src/boardbno055.cpp:192]: No need to change this. This file is actually for the Original Arduino Nano + BNO055 IMU nothing to to with the BLE33. You can remove all these changes in this file. (Pretty much old code now, although it still does connect and work with it. FYI https://github.com/dlktdr/HeadTracker/wiki/BNO055-Build-Instructions-(Discontinued)

BLE33 is all sent with the JSON settings and is dynamic. Should need to change anything here either but boardnanoble33.cpp is how it communicates with the board if you want to see.

File: gui/src/mainwindow.cpp (hunk 3 above)
Comment (dlktdr) [reply gui/src/mainwindow.cpp:776]: No nothing needed here. I only bother checking in the GUI/FW if the user selected two physical output pins that are the same. For the output channels you can select the same channel so whatever item is last in sense.cpp will be set. 
If I have alert=CH1 and tilt=CH1 will happily let you set this. Which is also why we should keep Tilt roll and pan as the last set item in the list.

File: firmware/src/src/sense.cpp (hunk 1 above)
Comment (dlktdr) [reply firmware/src/src/sense.cpp:444]: No you shouldn't use it. It is there to disable the TRP outputs if user has selected the long press mode to enable/disable those outputs. Was a feature request to make it more like the Trinity head tracker

File: firmware/src/src/sense.cpp [hunk 4]
```diff
@@ -433,13 +433,33 @@ void calculate_Thread()
...
+            }
+            if (sendingresetpulse) {
+                channel_data[alertch - 1] = TrackerSettings::MAX_PWM;
+                pulsetimer += (float)CALCULATE_PERIOD / 1000000.0;
+                if(pulsetimer > TrackerSettings::RECENTER_PULSE_DURATION) {
+                    sendingresetpulse = false;
+                }
+            }
+        }
+
```
Comment (dlktdr) [inline firmware/src/src/sense.cpp:462]: Looks good but can we move it above the setting of Tilt roll and Pan Line 433. I feel TRP should always take precedence if a user picks two of the same channels

**Comment by dlktdr** [comment]
> * what is process for updating online docs/help to explain this new field?

File: gui/src/mainwindow.ui [hunk 5]
```diff
@@ -6,8 +6,8 @@
    <rect>
     <x>0</x>
     <y>0</y>
-    <width>584</width>
-    <height>697</height>
+    <width>749</width>
+    <height>849</height>
```
Comment (dlktdr) [inline gui/src/mainwindow.ui:10]: We should leave this narrower width for users with small screens.

**Comment by dlktdr** [comment]
> so other things:
> 
> * what is the workflow for updating screen shots for README.md, docs, etc. -- I changed fields in the general tab of the gui
You would need to fork the wiki and make a PR to update the graphics in there. Or I can do it when released.
> * what is process for updating online docs/help to explain this new field?
Sent you a link so you can do edits on gitbook.
> * when it comes time to release, i see there are DBs of releases, etc. -- for now, perhaps @dlktdr can do this
I can take care of this. Before we release this & the other new items,
1) Fix debug in the GUI for sure. So you can just click the values you want to monitor. 
2) Gyro Calibration on boot needs more real world testing, it might be set to tightly to wait for zero. Also maybe something more obvious in the GUI too so the user knows there is no output because they haven't let it sit still.

File: firmware/src/src/sense.cpp (hunk 1 above)
Comment (gcmcnutt) [reply firmware/src/src/sense.cpp:444]: ok -- then my change is probably orthogonal -- i'll remove the todo comment

File: gui/src/boardbno055.cpp (hunk 2 above)
Comment (gcmcnutt) [reply gui/src/boardbno055.cpp:192]: ok, i'll revert the changes to this file.

File: firmware/src/src/sense.cpp (hunk 4 above)
Comment (gcmcnutt) [reply firmware/src/src/sense.cpp:462]: got it - this makes sense -- updated

File: gui/src/mainwindow.cpp (hunk 3 above)
Comment (gcmcnutt) [reply gui/src/mainwindow.cpp:776]: got it -- updated sense.cpp

File: gui/src/mainwindow.ui (hunk 5 above)
Comment (gcmcnutt) [reply gui/src/mainwindow.ui:10]: I am not comfortable with the way I added these items to the gui -- break the layout, add some items, then put in grid layout.

I will spend some time trying to undo as many of the changes as I can (in the xml directly) -- not sure what is best workflow here (the break layout seems pretty heavy handed)

**Comment by gcmcnutt** [comment]
> 

I did rebase my change on the prior commit as the system wasn't running that well to the GUI -- once I get closer, will rebase on the top of master and debug from there.

Question, this additional field (alertch) seems extensible, e.g. it should just work if someone updates firmware but doesn't change the saved settings.  But I'm not sure that upgrade pattern worked.  I'll need to test that case a bit more.

I am looking more at the mechanisms for sending data back to the unit (sort of a master/slave bidirectional thing, or s-bus, etc.  This feels like an OpenTX challenge, so I will dig a bit.

File: gui/src/mainwindow.ui [hunk 6]
```diff
@@ -24,11 +24,8 @@
     <normaloff>:/Icons/images/Icon256.png</normaloff>:/Icons/images/Icon256.png</iconset>
   </property>
   <widget class="QWidget" name="centralwidget">
-   <layout class="QVBoxLayout" name="verticalLayout_2" stretch="0,0,2">
-    <property name="sizeConstraint">
-     <enum>QLayout::SetDefaultConstraint</enum>
-    </property>
-    <item>
+   <layout class="QGridLayout" name="gridLayout_14">
```
Comment (gcmcnutt) [inline gui/src/mainwindow.ui:27]: yeah, this might be my bug, I may need to start again to get a Box Layout here...

File: gui/src/mainwindow.ui (hunk 5 above)
Comment (dlktdr) [reply gui/src/mainwindow.ui:10]: It can be difficult. I have had to hand edit in the past but I think you really need to do is expand the channel selector over one so it's spanning the whole row. When your editing the window just leave the width the minimum possible that everything fits.

Comment (dlktdr) [reply gui/src/mainwindow.ui:10]: > Question, this additional field (alertch) seems extensible, e.g. it should just work if someone updates firmware but doesn't change the saved settings. But I'm not sure that upgrade pattern worked. I'll need to test that case a bit more.

Yes it should just work if you update. On boot it is set default, if there is no JSON command saved in the flash it will stay default. I was trying to keep as much back/forwards compatibility as possible. So even an Old/Newer gui shouldn't crash at least. The way you did it works! The most recent Gryo changes I did on the other hand will cause some backwards issues.

> I am looking more at the mechanisms for sending data back to the unit (sort of a master/slave bidirectional thing, or s-bus, etc. This feels like an OpenTX challenge, so I will dig a bit.

Let me know if you want some info here. I've been working on https://github.com/EdgeTX/edgetx/pull/1374 and really need to get this one finished first. Wish I had more free time. the last few months have been crazy. 

The real challenge is that the current Frsky modules aren't bidirectional. Now that serial refactor is complete in ETX I would like to get back to working on the ESP as a open source option for communication bidirectionally.

Comment (gcmcnutt) [reply gui/src/mainwindow.ui:10]: Ok, a bit of hand edit and cleanup and the changes are easier to see in change review.

I'll push an update and then let me look at the other mechanisms, including your changes above.

I'm waiting on the in circuit debugger to arrive, plus it looks like I better plan a migrate to EdgeTx :-/

File: gui/src/mainwindow.ui (hunk 6 above)
Comment (gcmcnutt) [reply gui/src/mainwindow.ui:27]: yeah, this is cleaned up now

**Comment by gcmcnutt** [comment]
> > * what is the workflow for updating screen shots for README.md, docs, etc. -- I changed fields in the general tab of the gui
> 
> You would need to fork the wiki and make a PR to update the graphics in there. Or I can do it when released.

I was thinking of moving the README graphics into the source, vs being a reference out to media/gui -- that way updates track the version.  Does this look right or am I missing something here?


> > * what is process for updating online docs/help to explain this new field?
> 
> Sent you a link so you can do edits on gitbook.

Ok, this works, I will work on the updates here.

 
> > * when it comes time to release, i see there are DBs of releases, etc. -- for now, perhaps @dlktdr can do this
> 
> I can take care of this. Before we release this & the other new items,

ok

 
> 1. Fix debug in the GUI for sure. So you can just click the values you want to monitor.
> 2. Gyro Calibration on boot needs more real world testing, it might be set to tightly to wait for zero. Also maybe something more obvious in the GUI too so the user knows there is no output because they haven't let it sit still.

got it

**Comment by gcmcnutt** [comment]
I haven't rebased to the topmost changeset yet, as it wasn't working quite right -- so yeah, might be the IMU init stuff.  I'll take a look (I see tonight the debugger passed US customs, so could be soon now).

And I haven't changed the version yet -- still prelim.  Let me try out the rebase to top of master next.

**Comment by dlktdr** [comment]
This looks good to me, gave it a test works as it should, My radio now plays "Armed" when the camera is reset. Didn't have a Centered sound in there ;) 

Thx for the contribution keep the ideas coming. Appreciate the help.
